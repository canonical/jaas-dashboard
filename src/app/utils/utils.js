import { parseISO, formatDistanceToNow } from "date-fns";
import { useState } from "react";
import defaultCharmIcon from "static/images/icons/default-charm-icon.svg";

export function generateEntityIdentifier(charmId, name, subordinate) {
  if (!charmId) {
    return null;
  }

  return (
    <div className="entity-name u-truncate" title={name}>
      {subordinate && <span className="subordinate"></span>}
      {charmId && generateIconImg(name, charmId)}
      {name}
    </div>
  );
}

export const generateStatusElement = (
  status = "unknown",
  count,
  useIcon = true,
  actionsLogs = false
) => {
  let statusClass = status ? `is-${status.toLowerCase()}` : "";
  let countValue = "";
  if (count !== undefined) {
    countValue = ` (${count})`;
  }
  const className = useIcon ? "status-icon " + statusClass : "";
  // ActionLogs uses a spinner icon if 'running'
  if (actionsLogs && statusClass === "is-running") {
    return (
      <span className="status-icon is-spinner">
        <i className="p-icon--spinner u-animation--spin"></i>
        {status}
      </span>
    );
  }

  return (
    <span className={className}>
      {status}
      {countValue}
    </span>
  );
};

export const generateSpanClass = (className, value) => {
  return <span className={className}>{value}</span>;
};

// Highest status to the right
const statusOrder = ["running", "alert", "blocked"];

const setHighestStatus = (entityStatus, highestStatus) => {
  if (statusOrder.indexOf(entityStatus) > statusOrder.indexOf(highestStatus)) {
    return entityStatus;
  }
  return highestStatus;
};

// If it's the highest status then we don't need to continue looping
// applications or units.
const checkHighestStatus = (highestStatus) => {
  return highestStatus === statusOrder[statusOrder.length - 1];
};

export const getModelStatusGroupData = (model) => {
  let highestStatus = statusOrder[0]; // Set the highest status to the lowest.
  let messages = [];
  const applications = model.applications || {};
  Object.keys(applications).forEach((appName) => {
    const app = applications[appName];
    const { status: appStatus } = getApplicationStatusGroup(app);
    highestStatus = setHighestStatus(appStatus, highestStatus);
    if (checkHighestStatus(highestStatus)) {
      // If it's the highest status then we want to store the message.
      messages.push(app.status.info);
      return;
    }
    const units = app.units || {}; // subordinates do not have units.
    Object.keys(units).forEach((unitId) => {
      const unit = units[unitId];
      const { status: unitStatus } = getUnitStatusGroup(unit);
      highestStatus = setHighestStatus(unitStatus, highestStatus);
      if (checkHighestStatus(highestStatus)) {
        // If it's the highest status then we want to store the message.
        messages.push(unit["agent-status"].info);
        return;
      }
    });
  });
  return {
    highestStatus,
    messages,
  };
};

/**
  Returns the status for the application.
  @param {Object} application The application to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the application and any relevent messaging.
*/
export const getApplicationStatusGroup = (application) => {
  // Possible "blocked" or error states in application statuses.
  const blocked = ["blocked"];
  // Possible "alert" states in application statuses.
  const alert = ["unknown"];
  const status = application.status.status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns the status level for the machine.
  @param {Object} machine The machine to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the machine and any relevent messaging.
*/
export const getMachineStatusGroup = (machine) => {
  // Possible "blocked" or error states in machine statuses.
  const blocked = ["down"];
  // Possible "alert" states in machine statuses.
  const alert = ["pending"];
  const status = machine["agent-status"].status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns the status level for the unit.
  @param {Object} unit The unit to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the unit and any relevent messaging.
*/
export const getUnitStatusGroup = (unit) => {
  // Possible "blocked" or error states in the unit statuses.
  const blocked = ["lost"];
  // Possible "alert" states in the unit statuses.
  const alert = ["allocating"];
  const status = unit["agent-status"].status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns owner string from ownerTag
  @param {string} ownerTag The ownerTag identifier returns from the API
  @returns {string} The simplified owner string
*/
export const extractOwnerName = (tag) => {
  return tag.split("@")[0].replace("user-", "");
};

/**
  Pluralizes the supplied word based on the provided dataset.
  @param {Number} value The integer to be checked
  @param {string} string The item name to be pluralized
  @returns {string} The item pluralized if required
*/
export const pluralize = (value, string) => {
  const special = {
    active: "active",
    allocating: "allocating",
    down: "down",
    joined: "joined",
    lost: "lost",
    running: "running",
    started: "started",
    unknown: "unknown",
    waiting: "waiting",
  };
  if (value === 1) {
    return string;
  } else if (special[string]) {
    return special[string];
  }
  return `${string}s`;
};

/**
  Returns cloud string from cloudTag
  @param {string} cloudTag The cloudTag identifier returns from the API
  @returns {string} The simplified cloud string
*/
export const extractCloudName = (tag = "") => {
  return tag.replace("cloud-", "");
};

/**
  Returns credential string from Tag
  @param {string} cloudTag The cloudTag identifier returns from the API
  @returns {string} The simplified cloud string
*/
export const extractCredentialName = (tag) => {
  // @ is not there in local boostraps
  // cloudcred-localhost_admin_localhost
  if (!tag) return "-";
  let cred = tag.split("cloudcred-")[1];
  if (cred.indexOf("@") > -1) {
    return cred.split("@")[1].split("_")[1];
  }
  return cred.split("_")[1];
};

/**
  Returns the version of the supplied charm string.
  @param {String} charmName The full path of the charm e.g. cs:foo/bar-123
*/
export const extractRevisionNumber = (charmName = "") =>
  charmName.split("-").pop();

/**
  Returns a link to the charm icon for the provided charm name.
  @param {String} charmId The fully qualified charm name.
  @returns {String} The link to the charm icon.
*/
export const generateIconPath = (charmId) => {
  if (charmId.indexOf("local:") === 0) {
    return defaultCharmIcon;
  }
  if (charmId.indexOf("cs:") === 0) {
    // Strip unnessesary prefixes
    // before: cs:~containers/lxd-container-47
    // after:  containers/lxd-container-47
    charmId = charmId.replace("cs:~", "");
    charmId = charmId.replace("cs:", "");

    // Remove release from the charmId
    // before: containers/trusty/lxd-container-47
    // after:  containers/lxd-container-47
    charmId = charmId.replace("precise/", "");
    charmId = charmId.replace("trusty/", "");
    charmId = charmId.replace("xenial/", "");
    charmId = charmId.replace("bionic/", "");
    charmId = charmId.replace("focal/", "");

    // Combine owner and charm name
    // before: containers/lxd-container-47
    // after:  containers-lxd-container-47
    charmId = charmId.replace("/", "-");

    // Strip the revision number from the end
    // before: containers-lxd-container-47
    // after:  containers-lxd-container
    charmId = charmId.substr(0, charmId.lastIndexOf("-", charmId.length));

    return `https://charmhub.io/${charmId}/icon`;
  }
  if (charmId.indexOf("ch:") === 0) {
    // Regex explanation:
    // "ch:amd64/xenial/content-cache-425".match(/\/(.+)\/(.+)-\d+/)
    // Array(3) [ "/xenial/content-cache-425", "xenial", "content-cache" ]
    const charmName = charmId.match(/\/(.+)\/(.+)-\d+/)[2];
    return `https://charmhub.io/${charmName}/icon`;
  }
  return "";
};

/**
  @returns {Int || 0} Returns the current viewport width
*/
export const getViewportWidth = () => {
  const de = document.documentElement;
  return Math.max(de.clientWidth, window.innerWidth || 0);
};

/**
 * @param {function} fn Function to debounce
 * @param {Number} wait Time in milliseconds to wait
 */
export const debounce = (fn, wait) => {
  let t;
  return function () {
    clearTimeout(t);
    t = setTimeout(() => fn.apply(this, arguments), wait);
  };
};

/**
  Returns a bool on whether a user can admin controllers
  @param {Obj} conn The user connection.
  @returns {Boolean} If they are an admin or not.
*/
export const userIsControllerAdmin = (conn) => {
  const controllerAccess = conn?.info?.user?.controllerAccess;
  return ["superuser", "admin"].includes(controllerAccess);
};

export const isSet = (val) => val || val !== undefined;

export const extractRelationEndpoints = (relation) => {
  const endpoints = {};
  relation.endpoints.forEach((endpoint) => {
    const role = endpoint.relation.role;
    endpoints[role] =
      endpoint["application-name"] + ":" + endpoint.relation.name;
    endpoints[`${role}ApplicationName`] = endpoint["application-name"];
  });
  return endpoints;
};

const ImgWithFallback = ({
  src,
  fallback = defaultCharmIcon,
  alt,
  ...props
}) => {
  const [imgSrc, setImgSrc] = useState(src);
  const onError = () => setImgSrc(fallback);

  return (
    <img
      alt={alt}
      src={imgSrc ? imgSrc : fallback}
      onError={onError}
      {...props}
    />
  );
};

export const generateIconImg = (name, charmId) => {
  let iconSrc = defaultCharmIcon;
  if (charmId.indexOf("local:") !== 0) {
    iconSrc = generateIconPath(charmId);
  }
  return (
    <ImgWithFallback
      alt={name + " icon"}
      key={`${name}-${charmId}`}
      title={name}
      width="24"
      height="24"
      className="entity-icon"
      src={iconSrc}
    />
  );
};

export const generateRelationIconImage = (applicationName, applications) => {
  const application = applications[applicationName];
  if (!application || !applicationName) {
    return;
  }
  return generateIconImg(applicationName, application["charm-url"]);
};

export const formatFriendlyDateToNow = (date) => {
  const formattedDate = formatDistanceToNow(parseISO(date));
  return formattedDate.concat(" ago");
};

export const canAdministerModelAccess = (userName, modelUsers) => {
  let hasPermission = false;
  const sharingAccess = ["admin", "write", "owner"];
  modelUsers &&
    modelUsers.forEach((userObj) => {
      if (userObj.user === userName && sharingAccess.includes(userObj.access)) {
        hasPermission = true;
      }
    });
  return hasPermission;
};
