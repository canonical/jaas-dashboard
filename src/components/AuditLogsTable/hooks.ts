import { useCallback } from "react";
import { useSelector } from "react-redux";

import { useQueryParams } from "hooks/useQueryParams";
import {
  getControllerConnection,
  getWSControllerURL,
} from "store/general/selectors";
import { actions as jujuActions } from "store/juju";
import { getAuditEventsLimit } from "store/juju/selectors";
import { useAppDispatch, useAppSelector } from "store/store";

import type { AuditLogFilters } from "./AuditLogsTableFilters/AuditLogsTableFilters";
import { DEFAULT_AUDIT_LOG_FILTERS } from "./AuditLogsTableFilters/AuditLogsTableFilters";
import { DEFAULT_PAGE } from "./consts";

export const useFetchAuditEvents = () => {
  const dispatch = useAppDispatch();
  const wsControllerURL = useSelector(getWSControllerURL);
  const hasControllerConnection = useAppSelector((state) =>
    getControllerConnection(state, wsControllerURL)
  );

  const [queryParams] = useQueryParams<
    {
      page: string;
    } & AuditLogFilters
  >({
    page: DEFAULT_PAGE,
    ...DEFAULT_AUDIT_LOG_FILTERS,
  });
  const limit = Number(useSelector(getAuditEventsLimit));
  const page = Number(queryParams.page);
  return useCallback(() => {
    if (!wsControllerURL || !hasControllerConnection) {
      return;
    }
    dispatch(
      jujuActions.fetchAuditEvents({
        wsControllerURL,
        // Fetch an extra entry in order to check if there are more pages.
        limit: limit + 1,
        offset: (page - 1) * limit,
        // Pass undefined to the API if the filters haven't been set.
        after: queryParams.after
          ? new Date(queryParams.after).toISOString()
          : undefined,
        before: queryParams.before
          ? new Date(queryParams.before).toISOString()
          : undefined,
        // Convert the username to a user tag:
        "user-tag": queryParams.user ? `user-${queryParams.user}` : undefined,
        model: queryParams.model ?? undefined,
        method: queryParams.method ?? undefined,
      })
    );
  }, [
    dispatch,
    hasControllerConnection,
    limit,
    page,
    wsControllerURL,
    // Pass individual params so that this hook will only run if a specific
    // param changes as opposed to the queryParams object which will change
    // reference on each rerender.
    queryParams.after,
    queryParams.before,
    queryParams.user,
    queryParams.model,
    queryParams.method,
  ]);
};
