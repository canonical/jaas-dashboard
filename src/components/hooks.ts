import { useParams } from "react-router-dom";

import type { EntityDetailsRoute } from "components/Routes/Routes";

export const useEntityDetailsParams = () => {
  const { userName, modelName, appName, unitId, machineId } =
    useParams<EntityDetailsRoute>();
  return {
    appName,
    isNestedEntityPage: !!appName || !!unitId || !!machineId,
    machineId,
    modelName,
    unitId,
    userName,
  };
};
