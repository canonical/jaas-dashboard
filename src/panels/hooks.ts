import { useCallback } from "react";

import {
  type SetParams,
  type QueryParams,
  useQueryParams,
} from "hooks/useQueryParams";

export const usePanelQueryParams = <P extends QueryParams>(
  defaultQueryParams: P,
): [P, SetParams<P>, () => void] => {
  const [queryParams, setQueryParams] = useQueryParams<P>(defaultQueryParams);

  const handleRemovePanelQueryParams = useCallback(() => {
    const paramsToClear: Partial<P> = {};
    Object.keys(defaultQueryParams).forEach((param: keyof P) => {
      paramsToClear[param] = undefined;
    });
    setQueryParams(paramsToClear, { replace: true });
  }, [defaultQueryParams, setQueryParams]);

  return [queryParams, setQueryParams, handleRemovePanelQueryParams];
};
