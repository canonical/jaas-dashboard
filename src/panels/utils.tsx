import { useCallback } from "react";

import {
  type SetParams,
  type QueryParams,
  useQueryParams,
} from "hooks/useQueryParams";

type data = {
  th: string;
  td: string;
};

export const generatePanelTableRows = (data: data[]) => {
  return data.map(({ th, td }) => {
    return (
      <tr className="panel__tr" key={th + td}>
        <th className="panel__th">{th}</th>
        <td className="panel__td">{td}</td>
      </tr>
    );
  });
};

// TODO: Add tests for usePanelQueryParams.
export const usePanelQueryParams = <P extends QueryParams>(
  defaultQueryParams: P
): [P, SetParams<P>, () => void] => {
  const [queryParams, setQueryParams] = useQueryParams<P>(defaultQueryParams);

  const handleRemovePanelQueryParams = useCallback(() => {
    const paramsToClear: Partial<P> = {};
    Object.keys(defaultQueryParams).forEach((param: keyof P) => {
      paramsToClear[param] = undefined;
    });
    setQueryParams(paramsToClear);
  }, [defaultQueryParams, setQueryParams]);

  return [queryParams, setQueryParams, handleRemovePanelQueryParams];
};
