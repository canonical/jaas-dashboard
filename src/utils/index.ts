export { argPath } from "./argPath";
export { default as breakLines } from "./breakLines";
export { default as getMajorMinorVersion } from "./getMajorMinorVersion";
export { default as getUserName } from "./getUserName";
