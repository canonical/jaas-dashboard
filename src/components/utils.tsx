import classNames from "classnames";
import { parseISO, formatDistanceToNow } from "date-fns";
import { ImgHTMLAttributes, useState } from "react";
import defaultCharmIcon from "static/images/icons/default-charm-icon.svg";
import { generateIconPath } from "store/juju/utils/models";
import { ModelData } from "../juju/types";

export function generateEntityIdentifier(
  charmId: string,
  name: string,
  subordinate: boolean
) {
  if (!charmId) {
    return null;
  }

  return (
    <div className="entity-name u-truncate" title={name}>
      {subordinate && <span className="subordinate"></span>}
      {charmId && generateIconImg(name, charmId)}
      {name}
    </div>
  );
}

export const generateStatusElement = (
  status = "unknown",
  count?: number | null,
  useIcon = true,
  actionsLogs = false,
  className = null
) => {
  let statusClass = status ? `is-${status.toLowerCase()}` : "";
  let countValue = "";
  if (count || count === 0) {
    countValue = ` (${count})`;
  }
  const iconClasses = useIcon ? "status-icon " + statusClass : "";
  // ActionLogs uses a spinner icon if 'running'
  if (actionsLogs && statusClass === "is-running") {
    return (
      <span className="status-icon is-spinner">
        <i className="p-icon--spinner u-animation--spin"></i>
        {status}
      </span>
    );
  }

  return (
    <span className={classNames(iconClasses, className)}>
      {status}
      {countValue}
    </span>
  );
};

export const isSet = (val: unknown) => val || val !== undefined;

/**
  @returns {Int || 0} Returns the current viewport width
*/
export const getViewportWidth = () => {
  const de = document.documentElement;
  return Math.max(de.clientWidth, window.innerWidth || 0);
};

const ImgWithFallback = ({
  src,
  fallback = defaultCharmIcon,
  alt,
  ...props
}: {
  src: string;
  fallback?: string;
  alt: string;
} & Partial<ImgHTMLAttributes<HTMLImageElement>>) => {
  const [imgSrc, setImgSrc] = useState(src);
  const onError = () => setImgSrc(fallback);

  return (
    <img
      alt={alt}
      src={imgSrc ? imgSrc : fallback}
      onError={onError}
      {...props}
    />
  );
};

export const generateIconImg = (name: string, charmId: string) => {
  let iconSrc = defaultCharmIcon;
  if (charmId.indexOf("local:") !== 0) {
    iconSrc = generateIconPath(charmId);
  }
  return (
    <ImgWithFallback
      alt={name + " icon"}
      key={`${name}-${charmId}`}
      title={name}
      width="24"
      height="24"
      className="entity-icon"
      src={iconSrc}
    />
  );
};

export const generateRelationIconImage = (
  applicationName: string,
  applications: ModelData["applications"]
) => {
  const application = applications[applicationName];
  if (!application || !applicationName) {
    return;
  }
  return generateIconImg(applicationName, application["charm-url"]);
};

export const formatFriendlyDateToNow = (date: string) => {
  const formattedDate = formatDistanceToNow(parseISO(date));
  return formattedDate.concat(" ago");
};
