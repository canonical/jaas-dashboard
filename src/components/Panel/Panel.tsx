import type { PropsWithSpread } from "@canonical/react-components";
import { useListener } from "@canonical/react-components";
import classNames from "classnames";
import type { PropsWithChildren, ReactNode, MouseEvent } from "react";
import { forwardRef, useId } from "react";

import Aside from "components/Aside/Aside";
import type { Props as AsideProps } from "components/Aside/Aside";
import PanelHeader from "components/PanelHeader/PanelHeader";
import { useQueryParams } from "hooks/useQueryParams";

type Props = PropsWithSpread<
  {
    checkCanClose?: (e: KeyboardEvent | MouseEvent) => boolean;
    panelClassName: string;
    title: ReactNode;
  },
  PropsWithChildren & AsideProps
>;

export const close = {
  // Close panel if Escape key is pressed when panel active
  onEscape: (
    e: KeyboardEvent,
    queryStringSetter: (qs?: string | null) => void,
    checkCanClose?: Props["checkCanClose"]
  ) => {
    if (e.code === "Escape") {
      if (checkCanClose && !checkCanClose?.(e)) {
        return;
      }
      queryStringSetter(undefined);
    }
  },
  onClickOutside: (
    e: MouseEvent,
    queryStringSetter: (qs?: string | null) => void,
    checkCanClose?: Props["checkCanClose"]
  ) => {
    if (checkCanClose && !checkCanClose?.(e)) {
      return;
    }
    const target = e.target as HTMLElement;
    if (!target.closest(".p-panel")) {
      queryStringSetter(undefined);
    }
  },
};

export const Panel = forwardRef<HTMLDivElement, Props>(
  (
    { checkCanClose, children, panelClassName, title, ...props }: Props,
    ref
  ) => {
    const [, setPanelQs] = useQueryParams<{ panel: string | null }>({
      panel: null,
    });

    useListener(
      window,
      (e: KeyboardEvent) =>
        close.onEscape(
          e,
          () => setPanelQs(null, { replace: true }),
          checkCanClose
        ),
      "keydown"
    );
    useListener(
      window,
      (e: MouseEvent) =>
        close.onClickOutside(
          e,
          () => setPanelQs(null, { replace: true }),
          checkCanClose
        ),
      "click"
    );

    const titleId = useId();
    return (
      <Aside {...props} aria-labelledby={titleId} role="dialog">
        <div className={classNames("p-panel", panelClassName)} ref={ref}>
          <PanelHeader id={titleId} title={title} />
          {children}
        </div>
      </Aside>
    );
  }
);

export default Panel;
