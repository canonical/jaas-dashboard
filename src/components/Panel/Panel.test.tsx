import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { usePanelQueryParams } from "panels/hooks";
import { renderComponent } from "testing/utils";

import Panel from "./Panel";

describe("Panel", () => {
  const mockCheckCanClose = () => false;
  const mockRenderComponent = ({
    checkCanClose,
  }: {
    checkCanClose?: (
      e: KeyboardEvent | React.MouseEvent<Element, MouseEvent>
    ) => boolean;
  }) => {
    const MockPanelHeader = (): JSX.Element => {
      const [, , handleRemovePanelQueryParams] = usePanelQueryParams<{
        panel: null | string;
      }>({ panel: null });

      return (
        <Panel
          checkCanClose={checkCanClose}
          panelClassName="test-panel"
          title="Test panel"
          onRemovePanelQueryParams={handleRemovePanelQueryParams}
        >
          <>Test content</>
        </Panel>
      );
    };

    return renderComponent(<MockPanelHeader />, {
      url: "/foo?panel=share-model&externalParam=externalValue",
    });
  };

  it("should display the title and content", async () => {
    mockRenderComponent({});

    expect(screen.getByText("Test panel")).toBeInTheDocument();
    expect(screen.getByText("Test content")).toBeInTheDocument();
  });

  it("should give the panel a class name", async () => {
    mockRenderComponent({});

    expect(document.querySelector(".p-panel")).toHaveClass("test-panel");
  });

  it("should clear only the panel params when escape key is pressed", async () => {
    mockRenderComponent({});

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.type(window.document.documentElement, "{Escape}", {
      skipClick: true,
    });

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe("?externalParam=externalValue");
  });

  it("should not close if another key is pressed", async () => {
    mockRenderComponent({});

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.type(window.document.documentElement, "{Enter}", {
      skipClick: true,
    });

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );
  });

  it("should check if it can close when escape key is pressed", async () => {
    mockRenderComponent({ checkCanClose: mockCheckCanClose });

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.type(window.document.documentElement, "{Escape}", {
      skipClick: true,
    });

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );
  });

  it("should only clear the panel params when clicking outside the panel", async () => {
    mockRenderComponent({});

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.click(window.document.documentElement);

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe("?externalParam=externalValue");
  });

  it("should not clear the search params when clicking inside the panel", async () => {
    mockRenderComponent({});

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.click(screen.getByText("Test content"));

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );
  });

  it("should check if it can close when clicking outside the panel", async () => {
    mockRenderComponent({ checkCanClose: mockCheckCanClose });

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );

    await userEvent.click(window.document.documentElement);

    expect(window.location.pathname).toBe("/foo");
    expect(window.location.search).toBe(
      "?panel=share-model&externalParam=externalValue"
    );
  });
});
