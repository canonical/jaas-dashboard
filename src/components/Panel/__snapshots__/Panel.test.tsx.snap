// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel should render successfully 1`] = `
<div>
  <div
    aria-labelledby=":r0:"
    class="l-aside"
    role="dialog"
    style="transform: translateX(100%) translateZ(0);"
  >
    <div
      class="p-panel test-panel"
    >
      <div
        class="p-panel__header"
      >
        <div
          class="p-panel__title"
          id=":r0:"
        >
          Test panel
        </div>
        <div
          class="p-panel__controls"
        >
          <button
            class="p-button--base js-aside-close u-no-margin--bottom has-icon"
          >
            <i
              class="p-icon--close"
            />
          </button>
        </div>
      </div>
      Test content
    </div>
  </div>
</div>
`;
